### PR TITLE
[RW-5609][risk=no] Use environment variables for database configuration

### PR DESCRIPTION
**Pull request by dmohs**
The Google AppEngine runtime for Java 11 is substantially different than that for Java 8 and requires changes to how our application talks to Cloud SQL. As I iterated on this, having the additional burden of ambiguity around whether or not Spring would decide to use my specified configuration parameters would have caused significant fatigue. This PR addresses that problem by making that relationship more explicit.

How do I know that my changes to configuration parameters will affect the application?

Before:

```
% grep -R DB_CONNECTION_STRING
src/main/webapp/WEB-INF/appengine-web.xml.template
33:    <property name="spring.datasource.url" value="${DB_CONNECTION_STRING}"/>
37:    <property name="cdr.datasource.url" value="${CDR_DB_CONNECTION_STRING}"/>

db-cdr/vars.env
1:DB_CONNECTION_STRING=jdbc:mysql://db/cdr
```

```
% grep -R spring.datasource.url
src/main/webapp/WEB-INF/appengine-web.xml.template
33:    <property name="spring.datasource.url" value="${DB_CONNECTION_STRING}"/>

build.gradle
673:        'spring.datasource.url'              : "jdbc:mysql://${db_host}:${db_port}/workbench?useSSL=false",

src/test/resources/application.properties
2:spring.datasource.url=jdbc:h2:mem:db;DB_CLOSE_DELAY=-1;INIT=CREATE SCHEMA IF NOT EXISTS workbench;DB_CLOSE_ON_EXIT=FALSE
```
Now I must refer to Spring's documentation to determine how spring.datasource.url affects the system.

After:

```
% grep -R DB_HOST
src/main/java/org/pmiops/workbench/db/WorkbenchDbConfig.java
42:    // DB_HOST should be defined for local development, otherwise CLOUD_SQL_INSTANCE_NAME is
44:    Optional<String> dbHost = getEnv("DB_HOST");

vars.env
6:DB_HOST=127.0.0.1
```

Tested locally and deployed to AppEngine.

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [x] I have added explanatory comments where the logic is not obvious
- [x] I have run and tested this change locally, and my testing process is described here
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later
- [ ] If this includes an API change, I have run the E2E tests on this change against my local server with [yarn test-local](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples) because this PR won't be covered by the CircleCI tests 
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer in [Slack](https://pmi-engteam.slack.com/archives/C02MWP2RN5P)
- [ ] If this change impacts deployment safety (e.g. removing/altering APIs which are in use) I have documented these in the description
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and updated the appropriate API consumers
  * AoU UI
  * [Perf tests](https://github.com/broadinstitute/mcnulty/blob/develop/src/test/scala/services/AoU.scala)
  * [Researcher Directory export](https://github.com/all-of-us/workbench/wiki/Researcher-Directory-(RDR-export))
  * Cron tasks - for Offline*Controllers
  * SumoLogic - for EgressAlert

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/db/WorkbenchDbConfig.java [hunk 1]
```diff
@@ -36,14 +37,28 @@ public class WorkbenchDbConfig {
 
   @Primary
   @Bean(name = "dataSourceProperties")
-  @ConfigurationProperties(prefix = "spring.datasource")
   public DataSourceProperties dataSourceProperties() {
-    return new DataSourceProperties();
+    DataSourceProperties p = new DataSourceProperties();
+    // DB_HOST should be defined for local development, otherwise CLOUD_SQL_INSTANCE_NAME is
+    // required.
+    Optional<String> dbHost = getEnv("DB_HOST");
```
Comment (dmohs) [inline api/src/main/java/org/pmiops/workbench/db/WorkbenchDbConfig.java:44]: The right way to make this work for tests is to use a class here that can be replaced via dependency injection in tests.

**Review by evrii — APPROVED** [review]
This seems much more intuitive. i ran it locally using dev-up, and it worked successfully.

File: api/src/main/java/org/pmiops/workbench/db/WorkbenchDbConfig.java [hunk 2]
```diff
@@ -36,14 +37,28 @@ public class WorkbenchDbConfig {
 
   @Primary
   @Bean(name = "dataSourceProperties")
-  @ConfigurationProperties(prefix = "spring.datasource")
   public DataSourceProperties dataSourceProperties() {
-    return new DataSourceProperties();
+    DataSourceProperties p = new DataSourceProperties();
+    // DB_HOST should be defined for local development, otherwise CLOUD_SQL_INSTANCE_NAME is
```
[the remark this answers is not in the document]
Comment (dmohs) [reply api/src/main/java/org/pmiops/workbench/db/WorkbenchDbConfig.java:42]: This is how I'm handling it:
[api/src/main/java/org/pmiops/workbench/db/WorkbenchDbConfig.java](https://github.com/all-of-us/workbench/pull/7488/files#diff-afdac7a626bd3e7221effc8d5a279efef1a6418b7344fac438b1a3d338c76c17)

File: api/src/main/java/org/pmiops/workbench/db/WorkbenchDbConfig.java [hunk 3]
```diff
@@ -36,14 +37,28 @@ public class WorkbenchDbConfig {
...
+        dbHost.isPresent() ? "com.mysql.jdbc.Driver" : "com.mysql.jdbc.GoogleDriver");
+    String options = dbHost.isPresent() ? "useSSL=false" : "rewriteBatchedStatements=true";
+    p.setUrl(
+        String.format(
+            "jdbc:%smysql://%s/%s?%s",
+            dbHost.isPresent() ? "" : "google:",
+            dbHost.orElseGet(() -> getEnvRequired("CLOUD_SQL_INSTANCE_NAME")),
+            "workbench", // database name is consistent across environments
+            options));
+    p.setUsername("workbench"); // consistent across environments
```
[the remark this answers is not in the document]
Comment (dmohs) [reply api/src/main/java/org/pmiops/workbench/db/WorkbenchDbConfig.java:55]: Is there an environment where "workbench" is not the username?

**Pull request merged**